dts/bindings/base/base.yaml

```diff
@@ -60,6 +60,14 @@ properties:
     type: string-array
     description: name of each clock
 
+  assigned-clocks:
+    type: phandle-array
+    description: Assigned-clock information
+
+  assigned-clock-names:
+    type: string-array
+    description: Name of each assigned-clock
+
```

Comment:
Is this different from the `clocks` and `clock-names` property? seems this addition is a duplication

Reply:
Yes it is different. As I already mentioned in the commit description this came from Linux, see https://github.com/devicetree-org/dt-schema/blob/93ee8008959e362548168eaa7bdc20c115f3d586/dtschema/schemas/clock/clock.yaml#L126

I used this to give the developer an easy way to select the correct clock source to be used for some peripheral. There are cases that it complete changes the IP design between SoC's and this makes the glue possible with minimal implementation.

In the SAME54, the osc32kctrl is direct connected on the RTC peripheal. This means no generic clock associated. In this case, the clocks will use the osc32kctrl as direct source. The SAME54 uses MCLK peripheral to enable the peripheral.

```c
rtc: rtc@40002400 {
	compatible = "atmel,sam0-rtc";
	reg = <0x40002400 0x40>;
	interrupts = <11 0>;
	clock-generator = <0>;
	clocks = <&mclk 0x14 9 &osc32kctrl>;
	clock-names = "MCLK", "OSC32KCTRL";
	alarms-count = <2>;
	cal-constant = <(8192 * 128)>;
	assigned-clocks = <&osc32kctrl 0>;
	assigned-clock-names = "OSC32KCTRL";
	status = "disabled";
};
```

The above RTC is by default configured to use the ultra low power clock 32kHz. However the developer want to select the 1kHz from xtal. This means that instead redefine complete the clocks entry at board level he only need to override the **assigned-clock**.

```c
&rtc {
	assigned-clocks = <&osc32kctrl 4>;
};
```

Now, for the SAMR21 it is necessary to configure a generic clock and route it to RTC.

```c
&rtc {
	clocks = <&gclk 4>, <&pm 0x18 5>;
	clock-names = "GCLK", "PM";
	assigned-clocks = <&gclk 4>;
	assigned-clock-names = "GCLK";
};
```

In this case, I opted to already use the gclk[4] which is reserved by default at SoC initialization in Zephyr. Note that now user can select another generic clock line for these SoC's only changing the **assigned-clock**. Note that in this SoC the PM is used to enable the peripheral.

This means that developer can program another generic line and still assign to a peripheral using devicetree with almost no implementation. Yet the developer require to create the code that configure this generic line but soon it is possible to add the clock driver that do this job and then 100% of clocks can be handled by devicetree.

```c
&rtc {
	assigned-clocks = <&gclk x>;
};
```

To conclude, the **clocks** property assign what are the clocks associate with the peripheral. It allows to enable peripheral from PM or MCLK but do not route the clock to the peripheral because developer can complete define the lines. The **assigned-clocks** route the select line to finish the configuration. This means that developer do not touch clocks properties to develop and at board level he only select the clock lines.

Reply:
ok, so clocks is supported/connected clocks, and assigned-clocks are the ones of the connected clock which is actually used. Makes sense